### ISSUE TITLE
[JENKINS-4959] Block up-/downstream Projects of matrix projects

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/matrix-project-plugin#377]
<p>As implemented according to issue #1938 for maven builds, it should be possible<br/>
to block up-/downstream project in matrix builds too.</p>

<p>As a description, my problem:</p>

<p>We have matrix projects a b c and d. They build a dependency chain a-b-c-d.<br/>
At the time, if the projects changes the builds will start parallel and b,c and<br/>
d don't have the proper results of the upstream projects and fail.<br/>
If a is ready, b is scheduled and so on, the builds all are started again.</p>

<p>On the other side, if d is running and a changes a have to wait until d (or b or<br/>
c) is ready. At the time it will be started in parallel and probably the other<br/>
projects fail while loosing its dependencies.</p>

<p>Proposal:</p>

<p>Matrix builds should have a flag to be able to block the execution of a project<br/>
if an other project in the up-/downstream chain is running.</p>

<p>If the blocking project has finished, the Queue should be reordered according to<br/>
the up-/downstream chain and the projects should be build according to its<br/>
dependencies. (already queued projects should wait until they are triggered from<br/>
upstream project, or deleted from queue as they are started by the upstream<br/>
project again)</p>

<p>The parallel build of the axes should not be blocked.</p>

<p>BTW, this is NOT a duplicate of issue #1938, because matrix projects don't<br/>
support the blocking, like maven projects do, in using hudson version 1.335</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=wohauser">wohauser</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-4959" target="_blank">Block up-/downstream Projects of matrix projects</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: matrix-project-plugin
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2011-06-25T01:55:38+00:00
<li><b>votes</b>: 2
<li><b>watchers</b>: 1
<li><b>imported</b>: 2025-12-06
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>As implemented according to issue #1938 for maven builds, it should be possible
to block up-/downstream project in matrix builds too.</p>

<p>As a description, my problem:</p>

<p>We have matrix projects a b c and d. They build a dependency chain a-b-c-d.
At the time, if the projects changes the builds will start parallel and b,c and
d don't have the proper results of the upstream projects and fail.
If a is ready, b is scheduled and so on, the builds all are started again.</p>

<p>On the other side, if d is running and a changes a have to wait until d (or b or
c) is ready. At the time it will be started in parallel and probably the other
projects fail while loosing its dependencies.</p>

<p>Proposal:</p>

<p>Matrix builds should have a flag to be able to block the execution of a project
if an other project in the up-/downstream chain is running.</p>

<p>If the blocking project has finished, the Queue should be reordered according to
the up-/downstream chain and the projects should be build according to its
dependencies. (already queued projects should wait until they are triggered from
upstream project, or deleted from queue as they are started by the upstream
project again)</p>

<p>The parallel build of the axes should not be blocked.</p>

<p>BTW, this is NOT a duplicate of issue #1938, because matrix projects don't
support the blocking, like maven projects do, in using hudson version 1.335</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Platform: All, OS: All</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mbaker000:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/4959?focusedId=144706&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-144706">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-4959?focusedId=144706&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-144706">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I agree with this one, particularly with upstream jobs.  We have a build hierarchy where the later jobs need access to an artifact generated in a job higher in the job hierarchy.  The issue is that these later jobs take a while to run, and if one of the higher level jobs changes that artifact, using the lastSuccessfulBuild permalink doesn't work.</p>
</pre>
</details>

</ul>
</details>

<p>I agree with this one, particularly with upstream jobs.  We have a build hierarchy where the later jobs need access to an artifact generated in a job higher in the job hierarchy.  The issue is that these later jobs take a while to run, and if one of the higher level jobs changes that artifact, using the lastSuccessfulBuild permalink doesn't work.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>cjo9900:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/4959?focusedId=149813&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-149813">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-4959?focusedId=149813&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-149813">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Created a patch for this, with the latest release we should be able to just use the default handling for downstream blocking.
Added config option for this to matrix config properties.</p>

<p>Tested cases to make sure that it blocks correctly when a downstream job is running.</p>

<p>Pull request
<a href="https://github.com/jenkinsci/jenkins/pull/178" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/pull/178</a></p>
</pre>
</details>

</ul>
</details>

<p>Created a patch for this, with the latest release we should be able to just use the default handling for downstream blocking.<br/>
Added config option for this to matrix config properties.</p>

<p>Tested cases to make sure that it blocks correctly when a downstream job is running.</p>

<p>Pull request<br/>
<a href="https://github.com/jenkinsci/jenkins/pull/178" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/pull/178</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/4959?focusedId=149860&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-149860">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-4959?focusedId=149860&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-149860">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: cjo9900
Path:
 changelog.html
 core/src/main/resources/hudson/matrix/MatrixProject/configure-entries.jelly
<a href="http://jenkins-ci.org/commit/jenkins/591cbfd113cae53dd19c7c07119e97172fce6310" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/591cbfd113cae53dd19c7c07119e97172fce6310</a>
Log:
  <span class="error">&#91;FIXED JENKINS-4959&#93;</span> Add Downstream blocking for matrix projects.</p>



</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: cjo9900<br/>
Path:<br/>
 changelog.html<br/>
 core/src/main/resources/hudson/matrix/MatrixProject/configure-entries.jelly<br/>
<a href="http://jenkins-ci.org/commit/jenkins/591cbfd113cae53dd19c7c07119e97172fce6310" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/591cbfd113cae53dd19c7c07119e97172fce6310</a><br/>
Log:<br/>
  <span class="error">&#91;FIXED JENKINS-4959&#93;</span> Add Downstream blocking for matrix projects.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/dogfood.png" title="dogfood's avatar" /> dogfood:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/4959?focusedId=149911&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-149911">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-4959?focusedId=149911&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-149911">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Integrated in <span class="image-wrap" style=""><img src="http://ci.jenkins-ci.org/images/16x16/yellow.png" style="border: 0px solid black" /></span> <a href="http://ci.jenkins-ci.org/job/jenkins_main_trunk/914/" class="external-link" target="_blank" rel="nofollow noopener">jenkins_main_trunk #914</a></p>

</pre>
</details>

</ul>
</details>

<p>Integrated in <span class="image-wrap" style=""><img src="http://ci.jenkins-ci.org/images/16x16/yellow.png" style="border: 0px solid black" /></span> <a href="http://ci.jenkins-ci.org/job/jenkins_main_trunk/914/" class="external-link" target="_blank" rel="nofollow noopener">jenkins_main_trunk #914</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/4959?focusedId=152789&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-152789">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-4959?focusedId=152789&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-152789">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: cjo9900
Path:
 changelog.html
 core/src/main/resources/hudson/matrix/MatrixProject/configure-entries.jelly
<a href="http://jenkins-ci.org/commit/jenkins/591cbfd113cae53dd19c7c07119e97172fce6310" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/591cbfd113cae53dd19c7c07119e97172fce6310</a>
Log:
  <span class="error">&#91;FIXED JENKINS-4959&#93;</span> Add Downstream blocking for matrix projects.</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: cjo9900<br/>
Path:<br/>
 changelog.html<br/>
 core/src/main/resources/hudson/matrix/MatrixProject/configure-entries.jelly<br/>
<a href="http://jenkins-ci.org/commit/jenkins/591cbfd113cae53dd19c7c07119e97172fce6310" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/591cbfd113cae53dd19c7c07119e97172fce6310</a><br/>
Log:<br/>
  <span class="error">&#91;FIXED JENKINS-4959&#93;</span> Add Downstream blocking for matrix projects.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/4959?focusedId=154304&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-154304">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-4959?focusedId=154304&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-154304">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: cjo9900
Path:
 changelog.html
 core/src/main/resources/hudson/matrix/MatrixProject/configure-entries.jelly
<a href="http://jenkins-ci.org/commit/jenkins/591cbfd113cae53dd19c7c07119e97172fce6310" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/591cbfd113cae53dd19c7c07119e97172fce6310</a>
Log:
  <span class="error">&#91;FIXED JENKINS-4959&#93;</span> Add Downstream blocking for matrix projects.</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: cjo9900<br/>
Path:<br/>
 changelog.html<br/>
 core/src/main/resources/hudson/matrix/MatrixProject/configure-entries.jelly<br/>
<a href="http://jenkins-ci.org/commit/jenkins/591cbfd113cae53dd19c7c07119e97172fce6310" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/591cbfd113cae53dd19c7c07119e97172fce6310</a><br/>
Log:<br/>
  <span class="error">&#91;FIXED JENKINS-4959&#93;</span> Add Downstream blocking for matrix projects.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/4959?focusedId=163296&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-163296">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-4959?focusedId=163296&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-163296">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: cjo9900
Path:
 core/src/main/resources/hudson/matrix/MatrixProject/configure-entries.jelly
<a href="http://jenkins-ci.org/commit/matrix-project-plugin/afdeb3440c7c9e06e58d2825fd8292cfd8deb024" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/matrix-project-plugin/afdeb3440c7c9e06e58d2825fd8292cfd8deb024</a>
Log:
  <span class="error">&#91;FIXED JENKINS-4959&#93;</span> Add Downstream blocking for matrix projects.</p>

<p>Originally-Committed-As: 591cbfd113cae53dd19c7c07119e97172fce6310</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: cjo9900<br/>
Path:<br/>
 core/src/main/resources/hudson/matrix/MatrixProject/configure-entries.jelly<br/>
<a href="http://jenkins-ci.org/commit/matrix-project-plugin/afdeb3440c7c9e06e58d2825fd8292cfd8deb024" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/matrix-project-plugin/afdeb3440c7c9e06e58d2825fd8292cfd8deb024</a><br/>
Log:<br/>
  <span class="error">&#91;FIXED JENKINS-4959&#93;</span> Add Downstream blocking for matrix projects.</p>

<p>Originally-Committed-As: 591cbfd113cae53dd19c7c07119e97172fce6310</p>
